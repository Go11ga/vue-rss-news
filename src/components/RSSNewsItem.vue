<template>
    <div 
        class="rss-newsitem"
        :class="{ 'rss-newsitem--list' : getViewMode === 'list' }"
    >
        <div class="rss-newsitem__inner">

            <div class="rss-newsitem__header">
                <a :href="model.link" class="rss-newsitem__title" target="_blank">
                    {{ model.title }}
                </a>
            </div>

            <div class="rss-newsitem__content">
                <img v-if="model.image" class="rss-newsitem__img" :src="model.image" >
                <img v-else class="rss-newsitem__img" src="https://dummyimage.com/300x168/DCDCDC/fff">

                <div 
                    class="rss-newsitem__text"
                    :class="{ 'rss-newsitem__text--height-auto' : showHeightAuto }"
                >
                    {{ model.description }}
                </div>

                <button 
                    class="rss-newsitem__btn"
                    @click="toggleHeight"
                >
                    Подробнее
                </button>
            </div>

            <div class="rss-newsitem__footer">
                <a class="rss-newsitem__channel-link" :href="model.channelLink" target="_blank">
                    {{ model.channelLink | channelURL }}
                </a>
                <div class="rss-newsitem__date">
                    {{ model.pubDate | publicDate }}
                </div>
            </div>

        </div>
    </div>
</template>

<script>
import { mapGetters } from 'vuex'

export default {
    name: 'RSSNewsItem',

    props: {
        model: {
            type: Object,
            required: true
        }
    },

    data: () => ({
        showHeightAuto: false
    }),

    computed: {
        ...mapGetters([
            'getViewMode'
        ])
    },

    methods: {
        toggleHeight() {
            this.showHeightAuto = !this.showHeightAuto
        }
    },

    filters: {
        channelURL(url) {
            return url.replace(/https:\/\//, '')
        },

        publicDate(date) {
            let time = new Date(date)

            return `${time.toLocaleDateString()} ${time.toLocaleTimeString()}`
        }
    }
}
</script>
