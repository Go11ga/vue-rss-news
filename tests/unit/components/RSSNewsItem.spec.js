import { shallowMount, createLocalVue } from '@vue/test-utils'
import RSSNewsItem from '@/components/RSSNewsItem'
import Vuex from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('RSSNewsItem.vue', () => {
    let getters
    let store
    let model
    let wrapper

    beforeEach(() => {
        getters = {
            getViewMode: () => 'list'
        }

        store = new Vuex.Store({
            getters
        })

        model = {
            title: 'testTitle',
            image: 'testImage',
            description: 'testDescription',
            link: 'testLink',
            pubDate: 'testPubDate',
            channelLink: 'testChannelLink'
        }

        wrapper = shallowMount(RSSNewsItem, {
            localVue,
            store,
            propsData: {
                model
            }
        })
    })

    afterEach(() => {
        wrapper.destroy()
    })

    it('правильно рендерится', async () => {
        expect(wrapper.vm.$options.name).toMatch('RSSNewsItem')

        expect(wrapper.find('.rss-newsitem__img').attributes().src).toMatch('testImage')

        wrapper.setProps({
            model: {
                title: 'testTitle',
                image: null,
                description: 'testDescription',
                link: 'testLink',
                pubDate: 'testPubDate',
                channelLink: 'testChannelLink'
            }
        })

        await localVue.nextTick()

        expect(wrapper.find('.rss-newsitem__img').attributes().src).toMatch('https://dummyimage.com/300x168/DCDCDC/fff')
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    it('переключает режим отображения новости при клике на .rss-newsitem__btn', async () => {
        expect(wrapper.vm.$data.showHeightAuto).toBe(false)

        await wrapper.find('.rss-newsitem__btn').trigger('click')

        expect(wrapper.vm.$data.showHeightAuto).toBe(true)
        expect(wrapper.find('.rss-newsitem--list').classes()).toContain('rss-newsitem--list')
    })

    it('тест фильтра channelURL', () => {
        expect(RSSNewsItem.filters.channelURL('https://mos.ru')).toBe('mos.ru')
    })

    it('тест фильтра publicDate', () => {
        expect(RSSNewsItem.filters.publicDate('Tue, 19 Jul 2022 15:58:43 +0300')).toBe('19.07.2022 15:58:43')
    })
})
