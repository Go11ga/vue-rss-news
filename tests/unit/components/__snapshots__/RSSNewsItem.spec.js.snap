// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RSSNewsItem.vue snapshot 1`] = `
<div
  class="rss-newsitem rss-newsitem--list"
>
  <div
    class="rss-newsitem__inner"
  >
    <div
      class="rss-newsitem__header"
    >
      <a
        class="rss-newsitem__title"
        href="testLink"
        target="_blank"
      >
        
                    testTitle
                
      </a>
    </div>
     
    <div
      class="rss-newsitem__content"
    >
      <img
        class="rss-newsitem__img"
        src="testImage"
      />
       
      <div
        class="rss-newsitem__text"
      >
        
                    testDescription
                
      </div>
       
      <button
        class="rss-newsitem__btn"
      >
        
                    Подробнее
                
      </button>
    </div>
     
    <div
      class="rss-newsitem__footer"
    >
      <a
        class="rss-newsitem__channel-link"
        href="testChannelLink"
        target="_blank"
      >
        
                    testChannelLink
                
      </a>
       
      <div
        class="rss-newsitem__date"
      >
        
                    Invalid Date Invalid Date
                
      </div>
    </div>
  </div>
</div>
`;
